### ISSUE TITLE
Plugin-able Optimizers

### ISSUE THREAD

**Issue by FelixBenning** [tensorflow/tensorflow#49895]
**System information**
- Tensorflow version 2.5
- Are you willing to contribute it (Yes):

## Describe the feature and the current behavior/state.

Currently there are at least four different paths a user has to take to try out a different optimizer:

1. The Optimizer might be found under `tf.keras.optimizers`
2. If not, it might be included in `tfa.optimizers` (tensorflow/addons) or `tfp.optimizers` (tensorflow/probability)
3. If not, the user might need to (statically?) install a subclass `tf.keras.optimizers.Optimizer` defined in some github repo - if lucky there is a `setup.py`
4. There might not even be an implementation available subclassing `tf.keras.optimizers.Optimizer` so the user would have to reimplement that

Now wouldn't it be cool if any optimizer could be installed with `pip install my-optimizer` and then accessed using `tf.optimizers.MyOptimizer`? Now the python packaging guide includes a [section about plugins](https://packaging.python.org/guides/creating-and-discovering-plugins/#using-package-metadata) which would make just that possible:

Let's say I have implemented the class `MyOptimizer(tf.keras.optimizers.Optimizer)` in the file `my_optimizer.py`, then I could write the following `setup.py`:

```python
from setuptools import setup

setup(
    name='my-optimizer',
    version='1.0',
    py_modules=[ 'my_optimizer' ],
    install_requires= [...],
    entry_points='''
        [tf.optimizers]
        my-optimizer = my_optimizer:MyOptimizer
    '''
)
```

and upload the package. Now tensorflow could include this optimizer into `tf.optimizers` by doing the following:

```python
# tensorflow.optimizers.__init__.py

import importlib

PLUGINS = {
    plugin.attr : plugin.module for plugin in
    importlib.metadata.entry_points()["tf.optimizers"]
}

def __getattr__(name: str) -> Any:
    return getattr(
        importlib.import_module(PLUGINS[name]), 
        name
)
```

(Alternatively one could use importlib in the dictionary comprehension already for eager loading. I am not sure which one is better. But people probably only import Optimizers only once and do not use all of them so I defaulted to the lazy version)

This [works for python 3.7+](https://stackoverflow.com/a/48916205/6662425), for python 3.6 one would have to either [replace the module with a class](https://stackoverflow.com/a/7668273/6662425)  or use something like `setattr()` on the module, `exec(f"{plugin.attr} = importlib.import_module(plugin.module)")`, or some other magic

<details>
  <summary>Old (probably wrong) suggestion</summary>

I think I have misunderstood the purpose of the tf_export decorators here  

```python
import importlib
from tensorflow.python.util import tf_export
from importlib.metadata import entry_points

for plugin in entry_points()["tf.optimizers"]:
    custom_optim_class = getattr(importlib.import_module(plugin.module), plugin.attr)
    # export the plugin optimizer class under the name `plugin.attr` which is its self-given name 
    tf_export.tf_export("optimizers." + plugin.attr, metaclass=abc.ABCMeta)(custom_optim_class)
```

</details>

Users could then install any optimizer published with such an `entry_point` with pip and find the optimizer inside of `tf.optimizers`.

## Changes to the API

`tf.optimizers` would be filled with plugin Optimizers. The optimizers could alternatively be included into `tf.keras.optimizers` or any other place, ideally it would unify the place where one could find optimizers in the long term.

## Who will benefit from this feature?

I am assuming that the reason `tensorflow/addons` and `tensorflow/probability` exists, is to reduce the base install size. If all optimizers are plugins which are extremely easy to install but all found in the same module by some import magic, then it will be easier to try out custom optimizers. And if the current "batteries included" optimizers are converted into such plugins as well, then you could not only reduce the base install of tensorflow (making them [optional dependencies](https://setuptools.readthedocs.io/en/latest/userguide/dependency_management.html#optional-dependencies) - e.g. `tensorflow[all]` would still include them), but these default plugins would also become standalone "examples" how to subclass the Optimizer class properly.

It is also fairly trivial to create a GitHub Template Repository with such a setup.py and stumps of such an Optimizer class with included github actions that deploy the package to pypa.

So people who would want to create new optimizers could simply click "use template" on those repositories, replace "my-optimizer" occurances with their optimizer name, implement the method stumps, and set the pypa access token in the github secrets. And their optimizer would immediately be available to download with pip and automatically included in the `tf.optimizers` module.

This means that the first three install methods would be merged. And since creating a new optimizer is made easier using this template, it is likely that more people would do so, moving more of the fourth category optimizers into this as well.

The general idea is to take the "software engineering" aspect out of writing Optimizers as much as possible such that researchers have an easier time.

Lastly things like

```python
import tensorflow as tf

def optimizers_as_dict():
    return { name : cls for name, cls in tf.optimizers.__dict__.items() if issubclass(cls, tf.keras.optimizers.Optimizer) }
```

could provide easy dynamic access to all installed optimizers making it a lot easier to do AutoML with them. This could also be included as a method in `tf.optimizers`.

## Extension

This plugin approach could also be applied to layers, loss functions, etc. to make tensorflow more dynamic in general.

**Comment by FelixBenning** [comment]
Hm - circular imports might be a problem. Optimizers will need tensorflow, and tensorflow tries to import optimizers. But since optimizers will generally not use `tf.optimizers` this _should_ be manageable.

**Comment by FelixBenning** [comment]
Draft for such a Template: https://github.com/FelixBenning/tf-optimizer

**Comment by FelixBenning** [comment]
Demo Implementation for how tensorflow might incorporate such plugins into its modules https://github.com/FelixBenning/tensorflow-plugins (here the plugins are included into `tf_plugins.optimizers` etc.

I think the biggest issue will be static codecompletion. I am not sure if there is a way around this. At the moment this means that only the default objects are completed

**Comment by FelixBenning** [comment]
a different way to do this might be `namespaces packages` (https://docs.python.org/3/whatsnew/3.3.html#pep-420-implicit-namespace-packages) this would also allow people to write packages extending the `tensorflow.optimizers` This would require there to be no `__init__.py` file in `tensorflow.optimizers` not only in tensorflow itself, but also in none of the plugins.

The advantage of this approach is its simplicity (it would not require any import magic at all) and the fact that static code completion would probably work with it.

The disadvantage of this approach is the discipline it requires: If any of the plugins create an `__init__.py` in this folder things break. One could test whether that happends at import and warn the user - probably even tell them the culprit package but it is still a bit more annoying to deal with.

**Comment by qlzh727** [comment]
Adding Mihai from TF infra team and Tomer for addons.

**Comment by qlzh727** [comment]
Thanks for reporting the issue.

Firstly, if you want to access the optimizer you implemented (or implemented by someone else), having the optimizer under the tf.keras.optimizers name space doesn't actually change any functionality. Since TF has strong API contract, we actually moved most of the custom layers/optimizers/loss to the tf_addon repo, which suppose to be the location for your use case. The API boundary here is that anything under tf namespace will be ensure by the contract and backwards compatible. Having anything is not belonging to TF to be appear is defeating the purpose. 

Having tf_addons and other PIP package as the optional dependency is definitely something we want to avoid. TF is already a quite large package, and we are trying our best to reduce its size. We only include any dependency that is necessary. Also the TF -> Addon dependency seems to be a issue for circular dependency.

**Comment by FelixBenning** [comment]
@qlzh727 I know that this does not change the functionality of the optimizer - it just guarantees that *all* installed optimizers are found under the same namespace (also without having to send pull requests to tensorflow/addons and waiting for them to be processed). This might seem like a trivial change, but it is also a trivial change to implement and allows for easy looping over optimizers and easier use.

> Having tf_addons and other PIP package as the optional dependency is definitely something we want to avoid. TF is already a quite large package, and we are trying our best to reduce its size.

addons are already an optional dependency, that is the point of "optional". The point is, that you could have more fine grained control over what you want to install if every single optimizer is installed via pip. I just wanted to argue that you can still bundle some of these optimizers into optional dependencies. In fact this _reduces_ the size of tensorflow because optimizers, etc. are _only_ installed if you use these addon features

**Comment by qlzh727** [comment]
If we let all the custom optimizer co-exist in the same namespace, then we need to check for name conflict and loading priority, which could certainly cause confusing error when optimizer with same name appears. This will then led to the requirement that each optimizer need to have it unique name under the namespace, so TF has to have a way to somehow allow user to register that.

I don't think the proposed approach provides much benefit comparing to having it in its own PIP package and import with its own package/module path.

I would also suggest the addon repository again, which is the recommended repository for hosting this kind of contribution.
